We want to test whether the neural network [chemprop](https://github.com/chemprop/chemprop) performs well at our classification task. 
Running this on a seperate conda instance that has chemprop installed. 

In [ ]:
!cd .. && export PYTHONPATH=$PWD && chemprop_train --quiet --extra_metrics accuracy --data_path data/processed/smiles_to_receptor_one_hot.csv --dataset_type classification --save_dir models/chemprop_test/ 

/home/ian/anaconda3/envs/target-chemprop/lib/python3.8/site-packages/tap/tap.py:312: UserWarning: add_arguments is deprecated and will be removed on January 1st, 2021 please override "configure" instead and call add_argument there. 
  warn('add_arguments is deprecated and will be removed on January 1st, 2021'
423087it [00:18, 22791.93it/s]
 67%|█████████████████████▉           | 281429/423087 [01:36<00:46, 3066.39it/s][18:56:35] Conflicting single bond directions around double bond at index 5.
[18:56:35]   BondStereo set to STEREONONE and single bond directions set to NONE.
[18:56:35] Conflicting single bond directions around double bond at index 5.
[18:56:35]   BondStereo set to STEREONONE and single bond directions set to NONE.
[18:56:35] Conflicting single bond directions around double bond at index 5.
[18:56:35]   BondStereo set to STEREONONE and single bond directions set to NONE.
[18:56:35] Conflicting single bond directions around double bond at index 5.
[18:56:35]   BondStereo 

 10%|████                                    | 688/6770 [00:24<03:52, 26.12it/s][18:58:14] Conflicting single bond directions around double bond at index 5.
[18:58:14]   BondStereo set to STEREONONE and single bond directions set to NONE.

 24%|█████████▏                             | 1604/6770 [00:59<03:10, 27.05it/s]


 32%|████████████▋                          | 2195/6770 [01:22<02:04, 36.89it/s][18:59:12] Conflicting single bond directions around double bond at index 5.
[18:59:12]   BondStereo set to STEREONONE and single bond directions set to NONE.

 37%|██████████████▍                        | 2514/6770 [01:32<02:07, 33.41it/s]


 51%|███████████████████▉                   | 3454/6770 [02:08<04:43, 11.70it/s]


 56%|█████████████████████▉                 | 3811/6770 [02:19<01:40, 29.37it/s][19:00:10] Conflicting single bond directions around double bond at index 5.
[19:00:10]   BondStereo set to STEREONONE and single bond directions set to NONE.

 64%|█████████████████████████▏             | 4363/6770 [02:40<01:24, 28.59it/s]


 68%|██████████████████████████▋            | 4630/6770 [02:49<02:19, 15.34it/s][19:00:40] Conflicting single bond directions around double bond at index 5.
[19:00:40]   BondStereo set to STEREONONE and single bond directions set to NONE.

 76%|█████████████████████████████▋         | 5143/6770 [03:09<00:52, 30.84it/s][19:00:59] Conflicting single bond directions around double bond at index 5.
[19:00:59]   BondStereo set to STEREONONE and single bond directions set to NONE.

 78%|██████████████████████████████▍        | 5276/6770 [03:14<00:44, 33.38it/s]


 92%|███████████████████████████████████▊   | 6209/6770 [03:50<00:20, 27.55it/s]


100%|███████████████████████████████████████| 6770/6770 [04:11<00:00, 31.64it/s]
                                                                                
 65%|██████████████████████████▌              | 550/847 [00:13<00:07, 39.67it/s]


  2%|▋                                       | 125/6770 [00:04<03:52, 28.54it/s][19:02:43] Conflicting single bond directions around double bond at index 5.
[19:02:43]   BondStereo set to STEREONONE and single bond directions set to NONE.

  4%|█▋                                      | 293/6770 [00:10<03:47, 28.52it/s][19:02:49] Conflicting single bond directions around double bond at index 5.
[19:02:49]   BondStereo set to STEREONONE and single bond directions set to NONE.

 10%|████▏                                   | 708/6770 [00:26<03:31, 28.65it/s]


 24%|█████████▎                             | 1619/6770 [00:58<02:44, 31.25it/s]


 34%|█████████████▏                         | 2282/6770 [01:21<02:25, 30.74it/s][19:04:00] Conflicting single bond directions around double bond at index 5.
[19:04:00]   BondStereo set to STEREONONE and single bond directions set to NONE.

 37%|██████████████▌                        | 2534/6770 [01:32<02:35, 27.19it/s]


 45%|█████████████████▋                     | 3072/6770 [01:49<01:50, 33.35it/s][19:04:28] Conflicting single bond directions around double bond at index 6.
[19:04:28]   BondStereo set to STEREONONE and single bond directions set to NONE.

 51%|███████████████████▉                   | 3461/6770 [02:04<01:59, 27.60it/s]


 55%|█████████████████████▍                 | 3727/6770 [02:13<02:57, 17.13it/s][19:04:52] Conflicting single bond directions around double bond at index 5.
[19:04:52]   BondStereo set to STEREONONE and single bond directions set to NONE.

 56%|█████████████████████▊                 | 3783/6770 [02:17<06:56,  7.16it/s][19:04:56] Conflicting single bond directions around double bond at index 5.
[19:04:56]   BondStereo set to STEREONONE and single bond directions set to NONE.

 64%|█████████████████████████              | 4361/6770 [02:37<01:14, 32.35it/s][19:05:16] Conflicting single bond directions around double bond at index 5.
[19:05:16]   BondStereo set to STEREONONE and single bond directions set to NONE.



 78%|██████████████████████████████▍        | 5287/6770 [03:13<00:47, 31.35it/s]


 92%|███████████████████████████████████▊   | 6221/6770 [03:45<00:17, 31.77it/s]


100%|██████████████████████████████████████▉| 6769/6770 [04:06<00:00, 41.39it/s]
                                                                                
 66%|██████████████████████████▉              | 556/847 [00:18<00:06, 42.67it/s]


 11%|████▍                                   | 744/6770 [00:26<03:47, 26.48it/s]


 21%|████████▏                              | 1426/6770 [00:51<03:05, 28.76it/s][19:08:15] Conflicting single bond directions around double bond at index 5.
[19:08:15]   BondStereo set to STEREONONE and single bond directions set to NONE.

 24%|█████████▌                             | 1652/6770 [01:02<02:57, 28.86it/s]


 38%|██████████████▊                        | 2579/6770 [01:34<01:55, 36.22it/s]


 39%|███████████████▎                       | 2662/6770 [01:36<01:54, 35.73it/s][19:09:00] Conflicting single bond directions around double bond at index 5.
[19:09:00]   BondStereo set to STEREONONE and single bond directions set to NONE.

 40%|███████████████▌                       | 2705/6770 [01:38<01:53, 35.72it/s][19:09:01] Conflicting single bond directions around double bond at index 6.
[19:09:01]   BondStereo set to STEREONONE and single bond directions set to NONE.

 51%|████████████████████                   | 3473/6770 [02:09<04:06, 13.39it/s]


 63%|████████████████████████▍              | 4242/6770 [02:37<01:26, 29.20it/s][19:10:01] Conflicting single bond directions around double bond at index 5.
[19:10:01]   BondStereo set to STEREONONE and single bond directions set to NONE.

 65%|█████████████████████████▎             | 4402/6770 [02:43<01:27, 27.18it/s]


 79%|██████████████████████████████▋        | 5323/6770 [03:15<00:43, 33.19it/s]


 83%|████████████████████████████████▌      | 5645/6770 [03:29<00:32, 34.18it/s][19:10:53] Conflicting single bond directions around double bond at index 5.
[19:10:53]   BondStereo set to STEREONONE and single bond directions set to NONE.

 89%|██████████████████████████████████▌    | 5997/6770 [03:40<00:27, 28.04it/s][19:11:03] Conflicting single bond directions around double bond at index 5.
[19:11:03]   BondStereo set to STEREONONE and single bond directions set to NONE.

 92%|███████████████████████████████████▉   | 6232/6770 [03:51<00:16, 32.53it/s]


 94%|████████████████████████████████████▌  | 6344/6770 [03:54<00:17, 25.03it/s][19:11:18] Conflicting single bond directions around double bond at index 5.
[19:11:18]   BondStereo set to STEREONONE and single bond directions set to NONE.

100%|██████████████████████████████████████▉| 6768/6770 [04:18<00:00, 13.40it/s]
                                                                                
 42%|█████████████████▎                       | 357/847 [00:14<00:19, 25.41it/s]


  5%|█▉                                      | 319/6770 [00:16<05:05, 21.12it/s]


 16%|██████                                 | 1058/6770 [00:52<03:02, 31.32it/s]


 18%|██████▉                                | 1208/6770 [01:01<11:48,  7.85it/s][19:13:46] Conflicting single bond directions around double bond at index 5.
[19:13:46]   BondStereo set to STEREONONE and single bond directions set to NONE.

 28%|███████████                            | 1911/6770 [01:31<03:53, 20.83it/s]


 31%|████████████▏                          | 2113/6770 [01:38<02:30, 30.90it/s][19:14:24] Conflicting single bond directions around double bond at index 6.
[19:14:24]   BondStereo set to STEREONONE and single bond directions set to NONE.

 41%|████████████████                       | 2793/6770 [02:05<01:58, 33.57it/s]


 46%|█████████████████▊                     | 3095/6770 [02:15<02:17, 26.67it/s][19:15:01] Conflicting single bond directions around double bond at index 5.
[19:15:01]   BondStereo set to STEREONONE and single bond directions set to NONE.

 54%|█████████████████████▏                 | 3684/6770 [02:39<01:37, 31.55it/s]


 58%|██████████████████████▋                | 3936/6770 [02:52<01:31, 31.11it/s][19:15:37] Conflicting single bond directions around double bond at index 5.
[19:15:37]   BondStereo set to STEREONONE and single bond directions set to NONE.

 66%|█████████████████████████▊             | 4489/6770 [03:15<01:40, 22.70it/s][19:16:00] Conflicting single bond directions around double bond at index 5.
[19:16:00]   BondStereo set to STEREONONE and single bond directions set to NONE.

 67%|█████████████████████████▉             | 4506/6770 [03:17<05:11,  7.27it/s]


 69%|███████████████████████████            | 4693/6770 [03:23<01:15, 27.52it/s][19:16:09] Conflicting single bond directions around double bond at index 5.
[19:16:09]   BondStereo set to STEREONONE and single bond directions set to NONE.

 70%|███████████████████████████▏           | 4717/6770 [03:24<01:04, 31.87it/s][19:16:09] Conflicting single bond directions around double bond at index 5.
[19:16:09]   BondStereo set to STEREONONE and single bond directions set to NONE.

 79%|██████████████████████████████▊        | 5350/6770 [03:51<00:42, 33.11it/s]


 92%|███████████████████████████████████▊   | 6216/6770 [04:26<00:22, 24.35it/s]


100%|██████████████████████████████████████▉| 6769/6770 [04:48<00:00, 38.14it/s]
                                                                                
 53%|█████████████████████▋                   | 447/847 [00:17<00:10, 39.60it/s]


  9%|███▋                                    | 621/6770 [00:26<03:37, 28.28it/s]


 18%|███████                                | 1222/6770 [00:52<03:50, 24.09it/s][19:19:08] Conflicting single bond directions around double bond at index 5.
[19:19:08]   BondStereo set to STEREONONE and single bond directions set to NONE.

 22%|████████▌                              | 1488/6770 [01:01<07:18, 12.05it/s]


 28%|██████████▉                            | 1888/6770 [01:18<02:22, 34.17it/s][19:19:33] Conflicting single bond directions around double bond at index 5.
[19:19:33]   BondStereo set to STEREONONE and single bond directions set to NONE.

 35%|█████████████▌                         | 2363/6770 [01:38<02:30, 29.31it/s]


 49%|███████████████████                    | 3299/6770 [02:11<01:56, 29.73it/s]


 49%|███████████████████▏                   | 3321/6770 [02:12<01:39, 34.71it/s][19:20:28] Conflicting single bond directions around double bond at index 6.
[19:20:28]   BondStereo set to STEREONONE and single bond directions set to NONE.

 61%|███████████████████████▉               | 4159/6770 [02:48<02:33, 17.05it/s]


 68%|██████████████████████████▍            | 4595/6770 [03:11<01:54, 18.92it/s][19:21:27] Conflicting single bond directions around double bond at index 5.
[19:21:27]   BondStereo set to STEREONONE and single bond directions set to NONE.

 71%|███████████████████████████▊           | 4827/6770 [03:27<05:01,  6.44it/s]


 73%|████████████████████████████▌          | 4964/6770 [03:35<01:32, 19.57it/s][19:21:50] Conflicting single bond directions around double bond at index 5.
[19:21:50]   BondStereo set to STEREONONE and single bond directions set to NONE.

 77%|██████████████████████████████         | 5220/6770 [03:47<01:17, 19.90it/s][19:22:03] Conflicting single bond directions around double bond at index 5.
[19:22:03]   BondStereo set to STEREONONE and single bond directions set to NONE.
[19:22:03] Conflicting single bond directions around double bond at index 5.
[19:22:03]   BondStereo set to STEREONONE and single bond directions set to NONE.

 82%|███████████████████████████████▊       | 5519/6770 [04:06<01:46, 11.70it/s]


 92%|███████████████████████████████████▊   | 6206/6770 [04:43<00:18, 30.53it/s]


100%|██████████████████████████████████████▉| 6768/6770 [05:03<00:00, 39.91it/s]
                                                                                
 67%|███████████████████████████▎             | 565/847 [00:16<00:06, 41.59it/s]


  1%|▍                                        | 70/6770 [00:03<04:11, 26.66it/s][19:24:01] Conflicting single bond directions around double bond at index 5.
[19:24:01]   BondStereo set to STEREONONE and single bond directions set to NONE.

  7%|██▋                                     | 449/6770 [00:13<03:01, 34.74it/s][19:24:12] Conflicting single bond directions around double bond at index 5.
[19:24:12]   BondStereo set to STEREONONE and single bond directions set to NONE.

 11%|████▎                                   | 722/6770 [00:27<03:01, 33.35it/s]


 24%|█████████▌                             | 1658/6770 [01:00<02:25, 35.13it/s]


 26%|██████████▏                            | 1779/6770 [01:04<02:45, 30.09it/s][19:25:03] Conflicting single bond directions around double bond at index 5.
[19:25:03]   BondStereo set to STEREONONE and single bond directions set to NONE.

 30%|███████████▊                           | 2044/6770 [01:16<02:48, 27.98it/s][19:25:15] Conflicting single bond directions around double bond at index 5.
[19:25:15]   BondStereo set to STEREONONE and single bond directions set to NONE.

 38%|██████████████▊                        | 2563/6770 [01:35<03:16, 21.44it/s]


 52%|████████████████████▏                  | 3512/6770 [02:09<01:50, 29.60it/s]


 53%|████████████████████▍                  | 3555/6770 [02:10<01:35, 33.84it/s][19:26:09] Conflicting single bond directions around double bond at index 5.
[19:26:09]   BondStereo set to STEREONONE and single bond directions set to NONE.

 65%|█████████████████████████▌             | 4428/6770 [02:42<01:45, 22.22it/s]


 79%|██████████████████████████████▊        | 5340/6770 [03:17<00:35, 39.95it/s]


 92%|███████████████████████████████████▉   | 6240/6770 [03:49<00:14, 36.48it/s][19:27:48] Conflicting single bond directions around double bond at index 6.
[19:27:48]   BondStereo set to STEREONONE and single bond directions set to NONE.

 92%|████████████████████████████████████   | 6258/6770 [03:50<00:13, 37.04it/s]


 99%|██████████████████████████████████████▌| 6685/6770 [04:07<00:02, 34.09it/s][19:28:06] Conflicting single bond directions around double bond at index 5.
[19:28:06]   BondStereo set to STEREONONE and single bond directions set to NONE.

100%|██████████████████████████████████████▉| 6767/6770 [04:10<00:00, 35.81it/s]
                                                                                
 71%|█████████████████████████████            | 600/847 [00:19<00:05, 43.72it/s]


 93%|██████████████████████████████████████▏  | 790/847 [00:23<00:01, 41.45it/s]

Now that we have trained our neural net, let's see how it performed.

In [9]:
import pandas as pd

df = pd.read_csv("../models/chemprop_test/test_scores.csv")
df.head()

Task  Mean auc  \
0  11-beta-hydroxysteroid dehydrogenase 1, 11-bet...  0.998428   
1                  5-lipoxygenase activating protein  0.999995   
2  6-phosphofructo-2-kinase/fructose-2,6-bisphosp...  0.999979   
3                       ALK tyrosine kinase receptor  0.999047   
4  Acetyl-CoA carboxylase 2, Acetyl-CoA carboxyla...  0.999986   

   Standard deviation auc  Fold 0 auc  Mean accuracy  \
0                     0.0    0.998428       0.999102   
1                     0.0    0.999995       0.999858   
2                     0.0    0.999979       0.999953   
3                     0.0    0.999047       0.998771   
4                     0.0    0.999986       0.999811   

   Standard deviation accuracy  Fold 0 accuracy  
0                          0.0         0.999102  
1                          0.0         0.999858  
2                          0.0         0.999953  
3                          0.0         0.998771  
4                          0.0         0.999811

In [10]:
df.describe()

Mean auc  Standard deviation auc  Fold 0 auc  Mean accuracy  \
count  120.000000                   120.0  120.000000     120.000000   
mean     0.996506                     0.0    0.996506       0.997530   
std      0.004206                     0.0    0.004206       0.003403   
min      0.971807                     0.0    0.971807       0.982108   
25%      0.994863                     0.0    0.994863       0.997246   
50%      0.997647                     0.0    0.997647       0.998795   
75%      0.999528                     0.0    0.999528       0.999551   
max      1.000000                     0.0    1.000000       0.999976   

       Standard deviation accuracy  Fold 0 accuracy  
count                        120.0       120.000000  
mean                           0.0         0.997530  
std                            0.0         0.003403  
min                            0.0         0.982108  
25%                            0.0         0.997246  
50%                            0.0         0.998795  
75%                            0.0         0.999551  
max                            0.0         0.999976

Mean accuracy of 0.9975, with a minimum accuracy of 0.9821. Looks like chemprop performs very well. 

In [12]:
!cd .. && echo $'smiles\nCC(=O)OC1=CC=CC=C1C(=O)O\nO=C1NC2=CC=C([N+]([O-])=O)C=C2C(C3=CC=CC=C3F)=NC1' > test.csv

In [13]:
!cd .. && export PYTHONPATH=$PWD && chemprop_predict --test_path test.csv --checkpoint_dir models/chemprop_test/ --preds_path test_preds.csv

/home/ian/anaconda3/envs/target-chemprop/lib/python3.8/site-packages/tap/tap.py:312: UserWarning: add_arguments is deprecated and will be removed on January 1st, 2021 please override "configure" instead and call add_argument there. 
  warn('add_arguments is deprecated and will be removed on January 1st, 2021'
Loading training args
Loading data
2it [00:00, 13273.11it/s]
100%|██████████████████████████████████████████| 2/2 [00:00<00:00, 70492.50it/s]
Validating SMILES
Test size = 2
Predicting with an ensemble of 1 models
  0%|                                                     | 0/1 [00:00<?, ?it/s]Loading pretrained parameter "encoder.encoder.0.cached_zero_vector".
Loading pretrained parameter "encoder.encoder.0.W_i.weight".
Loading pretrained parameter "encoder.encoder.0.W_h.weight".
Loading pretrained parameter "encoder.encoder.0.W_o.weight".
Loading pretrained parameter "encoder.encoder.0.W_o.bias".
Loading pretrained parameter "ffn.1.weight".
Loading pretrained parameter "ffn.1.bia

In [14]:
df = pd.read_csv("../test_preds.csv")

df.set_index("smiles", inplace=True)

for _, row in df.iterrows():
    print(row.nlargest(5))
    print()

Cyclooxygenase-2, Cyclooxygenase-1                                                                                                                                                                                                                                                                               0.612759
Carbonic anhydrase II, Carbonic anhydrase VA, Carbonic anhydrase IX, Carbonic anhydrase I, Carbonic anhydrase IV, Carbonic anhydrase VB, Carbonic anhydrase VI, Carbonic anhydrase VII, Carbonic anhydrase XII, Carbonic anhydrase XIV, Carbonic anhydrase III, Carbonic anhydrase XIII, Carbonic anhydrase 1    0.376249
Histamine H1 receptor, Histamine H2 receptor, Histamine H3 receptor, Histamine H4 receptor                                                                                                                                                                                                                       0.013905
Estrogen receptor alpha, Estrogen receptor beta           

As we can see it identifies Asprin as acting on COX, and the experimental GABAergic NPS from the Predict NPS test set has GABA as the fourth most likely receptor. Not bad performance.

In [15]:
!cd .. && rm test.csv test_preds.csv